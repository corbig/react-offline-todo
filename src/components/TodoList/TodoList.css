.todo-list{
    width: 90%;
    margin: 20px;
}
.todo-done{
    text-decoration: line-through;
}