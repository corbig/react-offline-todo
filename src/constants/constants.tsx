/**
 * status of the todo
 */
export enum CodeStatus {
    TODO, DONE
}

/**
 * redux actions
 */
export const Action = {
    ADD_TODO: 'ADD_TODO',
    UPDATE_TODO: 'UPDATE_TODO',
    REMOVE_TODO: 'REMOVE_TODO',
    GET_ALL_TODOS: 'GET_ALL_TODOS'
};