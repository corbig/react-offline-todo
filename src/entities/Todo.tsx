import { TypedRecord } from 'typed-immutable-record';
import { CodeStatus } from '../constants/constants';

/** Immutable todo interface */
export interface TodoImmutable extends TypedRecord<TodoImmutable> {
    get(key: '_id'): string;
    get(key: '_rev'): string;
    get(key: 'content'): string;
    get(key: 'status'): CodeStatus;
    set(key: '_id', value: string): TodoImmutable;
    set(key: '_rev', value: string): TodoImmutable;
    set(key: 'content', value: string): TodoImmutable;
    set(key: 'status', value: CodeStatus): TodoImmutable;
}

/** Mutable todo interface */
export interface Todo {
    _id: string;  // id parameter of PouchDB document because here a todo is also a Pouch doc
    _rev?: string; // revision of the Pouch doc
    content: string; // content of the todo
    status: CodeStatus; // status of the todo
}