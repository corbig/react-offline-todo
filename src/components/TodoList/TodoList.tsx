import * as React from 'react';
import { connect } from 'react-redux';
import { changeStatus, removeTodo } from '../../actions/TodoActions';
import { TodoImmutable } from '../../entities/Todo';
import { List } from 'immutable';
import {
    List as ListComponent, ListItem, Checkbox, ListItemText,
    ListItemSecondaryAction, IconButton, Card
} from 'material-ui';
import { CodeStatus } from '../../constants/constants';
import DeleteIcon from 'material-ui-icons/Delete';
import './TodoList.css';

/**
 * Interfaces of TodoList state/props 
 */
interface TodoListState {
    
}

interface TodoListProps {
    changeStatus: (todo: TodoImmutable) => Promise<void>;
    removeTodo: (todo: TodoImmutable) => Promise<void>;
    todoList: List<TodoImmutable>;
}

/**
 * Function to call redux actions
 * @param dispatch redux function to notify the store
 */
const mapDispatchToProps = (dispatch: Function) => {
    return {
        changeStatus: (todo: TodoImmutable) => dispatch(changeStatus(todo)),
        removeTodo: (todo: TodoImmutable) => dispatch(removeTodo(todo))
    };
};

/**
 * Function to map redux store properties to TodoList properties
 * @param store 
 */
const mapStateToProps = (store: List<TodoImmutable>) => {
    return {
        todoList: store
    };
};

/**
 * TodoList Component
 */
export class TodoList extends React.Component<TodoListProps, TodoListState> {

    public render() {
        // if there is no todo we return an empty div
        if (this.props.todoList.size === 0) {
            return (<div />);
        }
        return (
            <Card className="todo-list">
                <ListComponent className="list-component">
                    {this.props.todoList.map((todo: TodoImmutable) => (
                        <ListItem
                            key={todo.get('id')}
                            dense={true}
                            button={true}
                            onClick={() => this.props.changeStatus(todo)}
                        >
                            <Checkbox
                                checked={todo.get('status') === CodeStatus.DONE}
                            />
                            <ListItemText
                                primary={todo.get('content')}
                                className={todo.get('status') === CodeStatus.DONE ? 'todo-done' : ''}
                            />
                            <ListItemSecondaryAction>
                                <IconButton onClick={() => this.props.removeTodo(todo)}>
                                    <DeleteIcon />
                                </IconButton>
                            </ListItemSecondaryAction>
                        </ListItem>
                    ))}
                </ListComponent>
            </Card>
        );
    }
}

/**
 * We connect the TodoList component to Redux
 */
export default connect(mapStateToProps, mapDispatchToProps)(TodoList);
