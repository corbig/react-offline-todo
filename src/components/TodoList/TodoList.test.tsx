import * as React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from './TodoList';
import { expect as chaiExpect } from 'chai';
import { IconButton, ListItem, Checkbox, ListItemText } from 'material-ui';
import * as sinon from 'sinon';
import { TodoImmutable, Todo } from '../../entities/Todo';
import { List, fromJS } from 'immutable';
import { CodeStatus } from '../../constants/constants';
import * as renderer from 'react-test-renderer';

describe('<TodoList />', () => {

    const changeStatus = (todo: TodoImmutable) => {
        return;
    };
    const removeTodo = (todo: TodoImmutable) => {
        return;
    };

    it('should render a <TodoList /> component', () => {

        const todoList: List<TodoImmutable> = fromJS([]);

        const wrapper = shallow(<TodoList todoList={todoList} changeStatus={changeStatus} removeTodo={removeTodo} />);

        chaiExpect(wrapper.find(ListItem).length).to.equal(0);

    });

    it('should render a <TodoList /> component with an unfinished todo', () => {

        const todo: Todo = {
            id: 'todo_1',
            content: 'my todo',
            status: CodeStatus.TODO
        };

        let todoList: List<TodoImmutable> = fromJS([todo]);

        const wrapper = shallow(<TodoList todoList={todoList} changeStatus={changeStatus} removeTodo={removeTodo} />);

        chaiExpect(wrapper.find(ListItem).length).to.equal(1);

        const todoElement = wrapper.find(ListItem);

        chaiExpect(todoElement.find(Checkbox).props().checked).to.equal(false);
        chaiExpect(todoElement.find(ListItemText).props().primary).to.equal('my todo');
        chaiExpect(todoElement.find(ListItemText).props().className).to.equal('');

    });

    it('should render a <TodoList /> component with an unfinished todo (Jest snapshot ver.)', () => {
        const todo: Todo = {
            id: 'todo_1',
            content: 'my todo',
            status: CodeStatus.TODO
        };

        let todoList: List<TodoImmutable> = fromJS([todo]);

        const componentTree = renderer.create(<TodoList todoList={todoList} changeStatus={changeStatus} removeTodo={removeTodo} />).toJSON();
        // Jest's expect function
        expect(componentTree).toMatchSnapshot();
    });

    it('should render a <TodoList /> component with a finished todo', () => {

        const todo: Todo = {
            id: 'todo_1',
            content: 'my todo',
            status: CodeStatus.DONE
        };

        let todoList: List<TodoImmutable> = fromJS([todo]);

        const wrapper = shallow(<TodoList todoList={todoList} changeStatus={changeStatus} removeTodo={removeTodo} />);

        chaiExpect(wrapper.find(ListItem).length).to.equal(1);

        const todoElement = wrapper.find(ListItem);

        chaiExpect(todoElement.find(Checkbox).props().checked).to.equal(true);
        chaiExpect(todoElement.find(ListItemText).props().primary).to.equal('my todo');
        chaiExpect(todoElement.find(ListItemText).props().className).to.equal('todo-done');

    });

    it('should render a <TodoList /> component with a finished todo (Jest snapshot ver.)', () => {
        const todo: Todo = {
            id: 'todo_1',
            content: 'my todo',
            status: CodeStatus.DONE
        };

        let todoList: List<TodoImmutable> = fromJS([todo]);

        const componentTree = renderer.create(<TodoList todoList={todoList} changeStatus={changeStatus} removeTodo={removeTodo} />).toJSON();
        
        // Jest's expect function
        expect(componentTree).toMatchSnapshot();
    });

    it('should render a <TodoList /> component and change todo status', () => {

        const changeStatusSpy = sinon.spy(changeStatus);

        const todo: Todo = {
            id: 'todo_1',
            content: 'my todo',
            status: CodeStatus.DONE
        };

        let todoList: List<TodoImmutable> = fromJS([todo]);

        const wrapper = shallow(<TodoList todoList={todoList} changeStatus={changeStatusSpy} removeTodo={removeTodo} />);

        chaiExpect(wrapper.find(ListItem).length).to.equal(1);

        const todoElement = wrapper.find(ListItem);

        todoElement.simulate('click');

        sinon.assert.calledOnce(changeStatusSpy);

    });

    it('should render a <TodoList /> component and remove todo', () => {

        const removeTodoSpy = sinon.spy(removeTodo);

        const todo: Todo = {
            id: 'todo_1',
            content: 'my todo',
            status: CodeStatus.DONE
        };

        let todoList: List<TodoImmutable> = fromJS([todo]);

        const wrapper = shallow(<TodoList todoList={todoList} changeStatus={changeStatus} removeTodo={removeTodoSpy} />);

        chaiExpect(wrapper.find(ListItem).length).to.equal(1);

        const removeButtonElement = wrapper.find(ListItem).find(IconButton);

        removeButtonElement.simulate('click');

        sinon.assert.calledOnce(removeTodoSpy);

    });

});