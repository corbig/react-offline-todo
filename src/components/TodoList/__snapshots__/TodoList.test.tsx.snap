// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> should render a <TodoList /> component with a finished todo (Jest snapshot ver.) 1`] = `
<div
  className="MuiPaper-root-1 MuiPaper-elevation2-5 MuiPaper-rounded-2 todo-list"
>
  <ul
    className="MuiList-root-28 MuiList-padding-29 list-component"
  >
    <li
      className="MuiListItem-container-33"
    >
      <div
        className="MuiButtonBase-root-42 MuiListItem-root-32 MuiListItem-dense-36 MuiListItem-gutters-39 MuiListItem-button-40 MuiListItem-secondaryAction-41"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButtonBase-root-42 MuiIconButton-root-54 MuiSwitchBase-root-49 MuiSwitchBase-default-51 MuiCheckbox-default-44 MuiSwitchBase-checked-52 MuiCheckbox-checked-45 MuiCheckbox-checkedSecondary-47"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role={undefined}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-59"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-60"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <input
              checked={true}
              className="MuiSwitchBase-input-50"
              disabled={undefined}
              id={undefined}
              name={undefined}
              onChange={[Function]}
              tabIndex={undefined}
              type="checkbox"
              value={undefined}
            />
          </span>
          <span
            className="MuiTouchRipple-root-66"
          />
        </span>
        <div
          className="MuiListItemText-root-73 MuiListItemText-dense-75 todo-done"
        >
          <h3
            className="MuiTypography-root-79 MuiTypography-subheading-86 MuiListItemText-primary-76 MuiListItemText-textDense-78"
          >
            my todo
          </h3>
        </div>
        <span
          className="MuiTouchRipple-root-66"
        />
      </div>
      <div
        className="MuiListItemSecondaryAction-root-103"
      >
        <button
          className="MuiButtonBase-root-42 MuiIconButton-root-54"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-59"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-60"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-66"
          />
        </button>
      </div>
    </li>
  </ul>
</div>
`;

exports[`<TodoList /> should render a <TodoList /> component with an unfinished todo (Jest snapshot ver.) 1`] = `
<div
  className="MuiPaper-root-1 MuiPaper-elevation2-5 MuiPaper-rounded-2 todo-list"
>
  <ul
    className="MuiList-root-28 MuiList-padding-29 list-component"
  >
    <li
      className="MuiListItem-container-33"
    >
      <div
        className="MuiButtonBase-root-42 MuiListItem-root-32 MuiListItem-dense-36 MuiListItem-gutters-39 MuiListItem-button-40 MuiListItem-secondaryAction-41"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButtonBase-root-42 MuiIconButton-root-54 MuiSwitchBase-root-49 MuiSwitchBase-default-51 MuiCheckbox-default-44"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role={undefined}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-59"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-60"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <input
              checked={false}
              className="MuiSwitchBase-input-50"
              disabled={undefined}
              id={undefined}
              name={undefined}
              onChange={[Function]}
              tabIndex={undefined}
              type="checkbox"
              value={undefined}
            />
          </span>
          <span
            className="MuiTouchRipple-root-66"
          />
        </span>
        <div
          className="MuiListItemText-root-73 MuiListItemText-dense-75"
        >
          <h3
            className="MuiTypography-root-79 MuiTypography-subheading-86 MuiListItemText-primary-76 MuiListItemText-textDense-78"
          >
            my todo
          </h3>
        </div>
        <span
          className="MuiTouchRipple-root-66"
        />
      </div>
      <div
        className="MuiListItemSecondaryAction-root-103"
      >
        <button
          className="MuiButtonBase-root-42 MuiIconButton-root-54"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-59"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-60"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-66"
          />
        </button>
      </div>
    </li>
  </ul>
</div>
`;
